uvicorn/supervisors/watchgodreload.py

```diff
@@ -12,23 +10,107 @@
 
 logger = logging.getLogger("uvicorn.error")
 
-if TYPE_CHECKING:
+if TYPE_CHECKING:  # pragma: no cover
+    import os
+
     DirEntry = os.DirEntry[str]
 
 
 class CustomWatcher(DefaultWatcher):
-    ignore_dotted_file_regex = r"^\/?(?:\w+\/)*(\.\w+)"
-    ignored: List[str] = []
+    def __init__(self, root_path: Path, config: Config):
+        default_includes = ["*.py"]
+        self.includes = [
+            default
+            for default in default_includes
+            if default not in config.reload_excludes
+        ]
+        self.includes.extend(config.reload_includes)
+        self.includes = list(set(self.includes))
+
+        default_excludes = [".*", ".py[cod]", ".sw.*", "~*"]
+        self.excludes = [
+            default
+            for default in default_excludes
+            if default not in config.reload_includes
+        ]
+        self.excludes.extend(config.reload_excludes)
+        self.excludes = list(set(self.excludes))
 
-    def __init__(self, root_path: str) -> None:
-        for t in self.ignored_file_regexes:
-            self.ignored.append(t)
-        self.ignored.append(self.ignore_dotted_file_regex)
-        self._ignored = tuple(re.compile(r) for r in self.ignored)
-        super().__init__(root_path)
+        self.watched_dirs: Dict[str, bool] = {}
+        self.watched_files: Dict[str, bool] = {}
+        self.dirs_includes = set(config.reload_dirs)
+        self.dirs_excludes = set(config.reload_dirs_excludes)
+        self.resolved_root = root_path
+        super().__init__(str(root_path))
 
     def should_watch_file(self, entry: "DirEntry") -> bool:
-        return not any(r.search(entry.name) for r in self._ignored)
+        cached_result = self.watched_files.get(entry.path)
+        if cached_result is not None:
+            return cached_result
```

Comment:
is it made to be faster ?

Reply:
Yes, depending on the amount of patterns it might take a while to match all the files. I had to implement it for the directories anyway and reused it for the files.
For the files the cache is something I am on the fence about...